server/src/main/java/org/elasticsearch/cluster/metadata/MetaDataIndexStateService.java

```diff
@@ -217,4 +223,121 @@ public ClusterState execute(ClusterState currentState) {
         });
     }
 
+    public void freezeIndex(final FreezeIndexClusterStateUpdateRequest request, final ActionListener<ClusterStateUpdateResponse> listener) {
+        if (request.indices() == null || request.indices().length == 0) {
+            throw new IllegalArgumentException("Index name is required");
+        }
+
+        final String indicesAsString = Arrays.toString(request.indices());
+        clusterService.submitStateUpdateTask("freeze-indices " + indicesAsString,
+                new AckedClusterStateUpdateTask<ClusterStateUpdateResponse>(Priority.URGENT, request, listener) {
+
+            @Override
+            protected ClusterStateUpdateResponse newResponse(boolean acknowledged) {
+                return new ClusterStateUpdateResponse(acknowledged);
+            }
+
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                Set<IndexMetaData> indicesToFreeze = new HashSet<>();
+                for (Index index : request.indices()) {
+                    final IndexMetaData indexMetaData = currentState.metaData().getIndexSafe(index);
+                    // Only open indices may be frozen
+                    if (indexMetaData.getState() == IndexMetaData.State.OPEN) {
+                        indicesToFreeze.add(indexMetaData);
+                    }
+                }
+
+                if (indicesToFreeze.isEmpty()) {
+                    return currentState;
+                }
+
+                // Check if index closing conflicts with any running restores
+                RestoreService.checkIndexClosing(currentState, indicesToFreeze);
+                // Check if index closing conflicts with any running snapshots
+                // TODO: we might not need to check this for freezing
+                SnapshotsService.checkIndexClosing(currentState, indicesToFreeze);
+                logger.info("freezing indices {}", indicesAsString);
+
+                MetaData.Builder mdBuilder = MetaData.builder(currentState.metaData());
+                ClusterBlocks.Builder blocksBuilder = ClusterBlocks.builder()
+                        .blocks(currentState.blocks());
```

Comment:
I see a lot of wrapping like this through the PR where it really would fit on one line. Would you please take advantage of the long lines that we have?